File: algorithms/crown_and_anchor.rb
# frozen_string_literal: true

faces = %i(
  crown
  anchor
  heart
  spade
  club
  diamond
).freeze

funds = 50
round = 0

while funds.positive? && funds < 100
  round += 1
  puts "第#{round}ラウンド"
  puts "  手持ち資金: ¥#{funds}"

  bettings_map = {}
  faces.each { |face| bettings_map[face] = 0 }

  total_bet = rand(1..funds)

  if total_bet == 7
    face = faces.sample
    bettings_map[face] += total_bet
  else
    remainings = total_bet
    while remainings.positive?
      bet = rand(1..remainings)
      face = faces.sample
      bettings_map[face] += bet
      remainings -= bet
    end
  end
  funds -= total_bet

  hand = []
  3.times { hand.push(faces.sample) }

  winnings = 0
  hand.each do |roll|
    winnings += bettings_map[roll]
  end

  puts "  掛け金: ¥#{total_bet} (" +
       bettings_map.map { |key, val| "#{key}: ¥#{val}" }.join(', ') +
       ')'

  puts "  出目: #{hand.join(', ')}"
  puts "  払戻金: ¥#{winnings}"

  funds += winnings
end

puts "  残金: ¥#{funds}"
puts 100 < funds ? "意気揚々と船に戻る。\n" : "トボトボと船に帰る。\n"

File: javascript/chapter4/crownAndAnchor.js
const FACES = ['crown', 'anchor', 'heart', 'spade', 'club', 'diamond'];

function rand(m, n) {
  return m + Math.floor((n - m + 1) * Math.random());
}

function randFace() {
  return FACES[rand(0, FACES.length -1)];
}

let funds = 50;
let round = 0

while (0 < funds && funds < 100) {
  round++;
  console.log(`第${round}ラウンド`);
  console.log(`  手持ち資金: ¥${funds}`);

  let bets = {
    crown: 0, anchor:  0,
    heart: 0, spade:   0,
    club:  0, diamond: 0,
  };
  let totalBet = rand(1, funds);

  if (totalBet === 7) {
    totalBet = funds;
    bets.heart = totalBet;
  } else {
    let remaining = totalBet;
    do {
      let bet = rand(1, remaining);
      let face = randFace();
      bets[face] = bets[face] + bet;
      remaining = remaining - bet;
    } while (0 < remaining);
  }
  funds = funds - totalBet

  console.log(`  賭け金: ¥${totalBet} (` +
    Object.keys(bets).map(face => `${face}: ¥${bets[face]}`).join(', ') +
    ')');

  const hand = [];
  for (let roll = 0; roll < 3; roll++) {
    hand.push(randFace());
  }
  console.log(`  出目: ${hand.join(', ')}`);

  let winnings = 0;
  for (let die = 0; die < hand.length; die++) {
    let face = hand[die];
    if (0 < bets[face]) winnings = winnings + bets[face];
  }
  funds = funds + winnings;
  console.log(`  払戻金: ¥${winnings}`)
}

console.log(`  残金: ¥${funds}`);
if (100 < funds) console.log('意気揚々と船に戻る。\n');
else console.log('トボトボと船に帰る。\n');
